File: apps/ai-service/app/ai/retrieval.py
from __future__ import annotations

from typing import Protocol

from app.ai.rerank import Reranker, score_to_distance
from app.ai.schemas import KBChunk, TicketContext
from app.ai.vectorstore import VectorStore


class Retriever(Protocol):
    def retrieve(self, ticket: TicketContext, k: int) -> list[KBChunk]: ...


def reciprocal_rank_fusion(
    ranked_lists: list[list[KBChunk]],
    *,
    rrf_k: int = 60,
) -> list[KBChunk]:
    """Fuse multiple ranked lists with Reciprocal Rank Fusion.

    ``score(doc) = Σ 1 / (rrf_k + rank)`` over lists where the doc appears
    (1-based ranks). Docs are keyed by ``id`` when present, else by
    ``(source, content)``. Returns chunks sorted by RRF score descending,
    with ``distance`` set to ``1 / (1 + rrf_score)`` (lower is better).
    """
    scores: dict[object, float] = {}
    best: dict[object, KBChunk] = {}

    for ranked in ranked_lists:
        for rank, chunk in enumerate(ranked, start=1):
            key = chunk.id if chunk.id is not None else (chunk.source, chunk.content)
            scores[key] = scores.get(key, 0.0) + 1.0 / (rrf_k + rank)
            # Prefer a copy that carries an id when available.
            prev = best.get(key)
            if prev is None or (prev.id is None and chunk.id is not None):
                best[key] = chunk

    fused = sorted(scores.items(), key=lambda item: item[1], reverse=True)
    out: list[KBChunk] = []
    for key, rrf_score in fused:
        chunk = best[key]
        out.append(
            KBChunk(
                id=chunk.id,
                content=chunk.content,
                source=chunk.source,
                distance=1.0 / (1.0 + rrf_score),
            )
        )
    return out


class VectorRetriever:
    """Retrieves knowledge-base passages via pure semantic (vector) search.

    A convenience wrapper for tests/simple callers. The triage agent instead
    calls the underlying primitives (``search_semantic``/``search_keyword`` +
    ``reciprocal_rank_fusion`` + the re-ranker) directly from its ``search_docs``
    and ``rerank_results`` tools.
    """

    def __init__(self, store: VectorStore) -> None:
        self.store = store

    def retrieve(self, ticket: TicketContext, k: int) -> list[KBChunk]:
        query = f"{ticket.title}\n\n{ticket.description}"
        return self.store.search(query, k)


class HybridRetriever:
    """Semantic + keyword retrieval → RRF fusion → CrossEncoder re-rank.

    Each lane fetches ``candidate_k`` hits; RRF merges them; the top
    ``rerank_pool`` fused docs are scored by the re-ranker; the best ``k`` are
    returned. Kept as a one-shot retriever for tests/simple callers; the triage
    agent runs the same pipeline in stages across its tool calls.
    """

    def __init__(
        self,
        store: VectorStore,
        reranker: Reranker,
        *,
        candidate_k: int = 20,
        rerank_pool: int = 10,
        rrf_k: int = 60,
    ) -> None:
        self.store = store
        self.reranker = reranker
        self.candidate_k = candidate_k
        self.rerank_pool = rerank_pool
        self.rrf_k = rrf_k

    def retrieve(self, ticket: TicketContext, k: int) -> list[KBChunk]:
        query = f"{ticket.title}\n\n{ticket.description}"
        semantic = self.store.search_semantic(query, self.candidate_k)
        keyword = self.store.search_keyword(query, self.candidate_k)
        fused = reciprocal_rank_fusion([semantic, keyword], rrf_k=self.rrf_k)
        if not fused:
            return []

        pool = fused[: self.rerank_pool]
        scores = self.reranker.score(query, [c.content for c in pool])
        scored = sorted(
            zip(pool, scores, strict=True),
            key=lambda item: item[1],
            reverse=True,
        )
        return [
            KBChunk(
                id=chunk.id,
                content=chunk.content,
                source=chunk.source,
                distance=score_to_distance(score),
            )
            for chunk, score in scored[:k]
        ]


def format_context(chunks: list[KBChunk]) -> str:
    """Render retrieved chunks into a grounded context block for the prompt."""
    if not chunks:
        return "(no relevant knowledge-base passages were found)"
    parts = []
    for i, chunk in enumerate(chunks, start=1):
        parts.append(f"[{i}] source={chunk.source}\n{chunk.content}")
    return "\n\n".join(parts)


def format_candidates(chunks: list[KBChunk]) -> str:
    """Render chunks for a tool result the model can cite.

    Labels each passage with its KB ``id`` (so the model can cite ``[id]`` and
    ``rerank_results`` can reference the same candidates); falls back to a 1-based
    index for chunks without an id (e.g. stubs).
    """
    if not chunks:
        return "(no relevant knowledge-base passages were found)"
    parts = []
    for i, chunk in enumerate(chunks, start=1):
        label = chunk.id if chunk.id is not None else i
        parts.append(
            f"[{label}] source={chunk.source} (distance={chunk.distance:.4f})\n{chunk.content}"
        )
    return "\n\n".join(parts)

File: apps/ai-service/internal/rag/retrieval.go
package rag

import (
	"context"
	"fmt"
	"sort"
	"strings"
)

// fusionKey identifies a chunk across ranked lists. Rows are keyed by their id;
// chunks without one (stubs, tests) fall back to source+content.
type fusionKey struct {
	id      int
	source  string
	content string
}

func keyFor(c KBChunk) fusionKey {
	if c.ID != 0 {
		return fusionKey{id: c.ID}
	}
	return fusionKey{source: c.Source, content: c.Content}
}

// ReciprocalRankFusion fuses multiple ranked lists with Reciprocal Rank Fusion.
//
//	score(doc) = Σ 1 / (rrfK + rank)
//
// over the lists where the doc appears (1-based ranks). Returns chunks sorted
// by RRF score descending — ties keep first-seen order — with Distance set to
// 1 / (1 + score) so lower-is-better still holds.
func ReciprocalRankFusion(rankedLists [][]KBChunk, rrfK int) []KBChunk {
	scores := make(map[fusionKey]float64)
	best := make(map[fusionKey]KBChunk)
	order := make([]fusionKey, 0)

	for _, ranked := range rankedLists {
		for i, chunk := range ranked {
			rank := i + 1
			key := keyFor(chunk)
			if _, seen := scores[key]; !seen {
				order = append(order, key)
			}
			scores[key] += 1.0 / float64(rrfK+rank)
			// Prefer a copy that carries an id when available.
			if prev, ok := best[key]; !ok || (prev.ID == 0 && chunk.ID != 0) {
				best[key] = chunk
			}
		}
	}

	sort.SliceStable(order, func(i, j int) bool {
		return scores[order[i]] > scores[order[j]]
	})

	out := make([]KBChunk, 0, len(order))
	for _, key := range order {
		chunk := best[key]
		out = append(out, KBChunk{
			ID:       chunk.ID,
			Content:  chunk.Content,
			Source:   chunk.Source,
			Distance: 1.0 / (1.0 + scores[key]),
		})
	}
	return out
}

// Retriever fetches knowledge-base passages relevant to a free-text query.
type Retriever interface {
	Retrieve(ctx context.Context, query string, k int) ([]KBChunk, error)
}

// VectorRetriever retrieves knowledge-base passages via pure semantic (vector)
// search.
//
// A convenience wrapper for tests and simple callers. The triage agent instead
// drives the underlying primitives (SearchSemantic/SearchKeyword +
// ReciprocalRankFusion + the re-ranker) from its search_docs and
// rerank_results tools.
type VectorRetriever struct{ Store *VectorStore }

func (v *VectorRetriever) Retrieve(ctx context.Context, query string, k int) ([]KBChunk, error) {
	return v.Store.SearchSemantic(ctx, query, k)
}

// HybridRetriever runs semantic + keyword retrieval, fuses with RRF, then
// re-ranks with a cross-encoder.
//
// Each lane fetches CandidateK hits; RRF merges them; the top RerankPool fused
// docs are scored by the re-ranker; the best k are returned. Kept as a one-shot
// retriever for tests and simple callers; the triage agent runs the same
// pipeline in stages across its tool calls.
type HybridRetriever struct {
	Store      SearchStore
	Reranker   Reranker
	CandidateK int
	RerankPool int
	RRFK       int
}

// SearchStore is the retrieval surface HybridRetriever needs (satisfied by
// *VectorStore; stubbed in tests).
type SearchStore interface {
	SearchSemantic(ctx context.Context, query string, k int) ([]KBChunk, error)
	SearchKeyword(ctx context.Context, query string, k int) ([]KBChunk, error)
}

func (h *HybridRetriever) Retrieve(ctx context.Context, query string, k int) ([]KBChunk, error) {
	semantic, err := h.Store.SearchSemantic(ctx, query, h.CandidateK)
	if err != nil {
		return nil, err
	}
	keyword, err := h.Store.SearchKeyword(ctx, query, h.CandidateK)
	if err != nil {
		return nil, err
	}
	fused := ReciprocalRankFusion([][]KBChunk{semantic, keyword}, h.RRFK)
	if len(fused) == 0 {
		return nil, nil
	}

	pool := fused
	if len(pool) > h.RerankPool {
		pool = pool[:h.RerankPool]
	}
	scored, err := h.rerank(ctx, query, pool)
	if err != nil {
		return nil, err
	}
	if len(scored) > k {
		scored = scored[:k]
	}
	return scored, nil
}

// rerank scores pool against query and returns it best-first, with Distance
// rewritten from the re-ranker score.
func (h *HybridRetriever) rerank(ctx context.Context, query string, pool []KBChunk) ([]KBChunk, error) {
	passages := make([]string, len(pool))
	for i, c := range pool {
		passages[i] = c.Content
	}
	scores, err := h.Reranker.Score(ctx, query, passages)
	if err != nil {
		return nil, err
	}
	if len(scores) != len(pool) {
		return nil, fmt.Errorf("reranker returned %d scores for %d passages", len(scores), len(pool))
	}

	idx := make([]int, len(pool))
	for i := range idx {
		idx[i] = i
	}
	sort.SliceStable(idx, func(a, b int) bool { return scores[idx[a]] > scores[idx[b]] })

	out := make([]KBChunk, len(pool))
	for rank, i := range idx {
		out[rank] = KBChunk{
			ID:       pool[i].ID,
			Content:  pool[i].Content,
			Source:   pool[i].Source,
			Distance: ScoreToDistance(scores[i]),
		}
	}
	return out, nil
}

const noPassages = "(no relevant knowledge-base passages were found)"

// FormatContext renders retrieved chunks into a grounded context block for the
// prompt.
func FormatContext(chunks []KBChunk) string {
	if len(chunks) == 0 {
		return noPassages
	}
	parts := make([]string, len(chunks))
	for i, chunk := range chunks {
		parts[i] = fmt.Sprintf("[%d] source=%s\n%s", i+1, chunk.Source, chunk.Content)
	}
	return strings.Join(parts, "\n\n")
}

// FormatCandidates renders chunks for a tool result the model can cite.
//
// Each passage is labelled with its KB id (so the model can cite "[id]" and
// rerank_results can reference the same candidates); chunks without an id
// (e.g. stubs) fall back to a 1-based index.
func FormatCandidates(chunks []KBChunk) string {
	if len(chunks) == 0 {
		return noPassages
	}
	parts := make([]string, len(chunks))
	for i, chunk := range chunks {
		label := chunk.ID
		if label == 0 {
			label = i + 1
		}
		parts[i] = fmt.Sprintf("[%d] source=%s (distance=%.4f)\n%s",
			label, chunk.Source, chunk.Distance, chunk.Content)
	}
	return strings.Join(parts, "\n\n")
}
